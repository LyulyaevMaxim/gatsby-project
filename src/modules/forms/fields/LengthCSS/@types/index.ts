//namespace NFieldLengthCSS
export enum defaultUnitsOfMeasure {
  'auto' = '-',
  // 'initial' = '-',
  // 'inherit' = '-',
}

export enum unitsOfLengthMeasure {
  'cap' = 'cap',
  'ch' = 'ch',
  'em' = 'em',
  'ex' = 'ex',
  'ic' = 'ic',
  'lh' = 'lh',
  'rem' = 'rem',
  'rlh' = 'rlh',
  'vh' = 'vh',
  'vw' = 'vw',
  'vi' = 'vi',
  'vb' = 'vb',
  'vmin' = 'vmin',
  'vmax' = 'vmax',
  'px' = 'px',
  'cm' = 'cm',
  'mm' = 'mm',
  'Q' = 'Q',
  'in' = 'in',
  'pc' = 'pc',
  'pt' = 'pt',
}

export enum additionalUnitsOfMeasure {
  '%' = '%',
}

export const numericUnits = {
    ...unitsOfLengthMeasure,
    ...additionalUnitsOfMeasure,
  },
  allUnits = {
    ...defaultUnitsOfMeasure,
    ...numericUnits,
  }

export type allUnits =
  | keyof typeof defaultUnitsOfMeasure
  | keyof typeof unitsOfLengthMeasure
  | keyof typeof additionalUnitsOfMeasure

export interface IFieldLengthCSS {
  ['menu-items']?: Array<allUnits>
  value?: string
  onChange?: (newState: { value: string; unit: allUnits }) => any
}

export interface IFieldInputLengthCSS {
  value: string
  setParentState?: (valuesFromInput: IPatternCorrectInputByGroups) => void
}

export interface IFieldSelectLengthCSS {
  menuItems: Array<allUnits>
  value: allUnits
  setParentState?: (valuesFromSelect: { unit: allUnits }) => void
}

export interface IPatternCorrectInputByGroups {
  defaultValue?: string
  number?: string
  unit?: allUnits
}
