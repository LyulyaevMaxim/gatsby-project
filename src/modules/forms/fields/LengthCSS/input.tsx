import React from 'react'
import * as I from './types'

interface IInputState {
  value: string
  isValid: null | boolean
}

export const FieldInputLengthCSS: React.FC<I.IFieldInputLengthCSS> = React.memo(props => {
  const [input, setInput] = React.useState<IInputState>({
      value: props.value,
      isValid: props.value ? patterns.correctInput.test(props.value) : null,
    }),
    inputOnChange = React.useCallback(event => {
      const elValue = event.target.value
      setInput(state => ({ ...state, value: elValue, isValid: patterns.similarInput.test(elValue) }))
    }, []),
    inputOnBlur = React.useCallback(event => {
      const elValue = event.target.value,
        matches = patterns.correctInput.exec(elValue)
      let newState = { isValid: false, value: elValue }

      if (matches) {
        const { defaultValue, number, unit }: I.IPatternCorrectInputByGroups = matches.groups || {}
        newState = { isValid: true, value: number || defaultValue }
        if (typeof props.setParentState === 'function') props.setParentState({ defaultValue, number, unit })
      }
      setInput(state => ({ ...state, ...newState }))
    }, [])

  //prop 'value' was updated
  React.useEffect(() => {
    if (props.value !== input.value) setInput(state => ({ ...state, isValid: true, value: props.value }))
  }, [props.value])

  return (
    <input
      data-testid="field-lengthCSS-input"
      value={input.value}
      onChange={inputOnChange}
      onBlur={inputOnBlur}
      type="text"
      css={[styles.input, input.value && styles[input.isValid ? 'valid' : 'invalid']]}
    />
  )
})

const styles = {
  input: `
    border: 1px solid;
    &:focus {
      outline: none;
    }
  `,
  valid: `
    border-color: green;
  `,
  invalid: `
    border-color: red;
  `,
}

export const patterns = (() => {
  const defaultValues = `(?<defaultValue>${Object.keys(I.defaultUnitsOfMeasure).join('|')})`,
    //similar to correct: 1, 1., 1.1, 1p, 1ppp
    numbersWithLetters = `[-]?\\d+((\\.?(\\d+[a-z%]*)?)|[a-z%]*)`,
    //only correct: 1, 1.1, 1px
    numbersWithUnits = `((?<number>[-]?(\\d*\\.)?\\d+)(?<unit>${Object.keys(I.numericUnits).join('|')})?)`

  return {
    similarInput: new RegExp(`^([a-z]+|${numbersWithLetters})$`),
    correctInput: new RegExp(`^(${defaultValues}|${numbersWithUnits})$`),
  }
})()
