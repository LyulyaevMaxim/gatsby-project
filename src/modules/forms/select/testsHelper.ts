import { fireEvent, RenderResult } from '@testing-library/react'

interface ICreateSelectInterface {
  container: RenderResult['container']
}

export const createSelectInterface = ({ container }: ICreateSelectInterface) => {
  return {
    get value() {
      return (container.querySelector('[name=field-lengthCSS-select]') as any).value
    },
    get options() {
      if (!this.isMenuOpen) this.openMenu()
      return [...(container.querySelector('.select__menu-list') as any).children]
    },
    get input() {
      return container.querySelector('.select__input input')
    },
    get control() {
      return container.querySelector('.select__control')
    },
    get isMenuOpen() {
      return [...(this.control as any).classList].includes('select__control--menu-is-open')
    },
    openMenu() {
      fireEvent.focus(this.input as any)
      fireEvent.mouseDown(this.control as any)
    },
    getOption(optionValue: string) {
      return this.options.find((option: HTMLElement) => option.textContent === optionValue)
    },
    changeOption(newOption: string) {
      if (!this.isMenuOpen) this.openMenu()
      fireEvent.click(this.getOption(newOption))
    },
  }
}
