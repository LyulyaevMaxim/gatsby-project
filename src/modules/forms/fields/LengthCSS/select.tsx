import React from 'react'
import { Select } from 'modules/forms/select'
import * as NFieldLengthCSS from './@types'

export const FieldSelectLengthCSS: React.FC<NFieldLengthCSS.IFieldSelectLengthCSS> = React.memo(props => {
  const { menuItems } = props,
    options = React.useMemo(
      () =>
        ([...menuItems, !menuItems.includes(props.value) && props.value].filter(Boolean) as Array<
          NFieldLengthCSS.allUnits
        >).map(currentOption => ({ value: currentOption, label: NFieldLengthCSS.allUnits[currentOption] })),
      [menuItems, props.value]
    ),
    onChange = React.useCallback(newSelectedOption => {
      if (props.setParentState) props.setParentState({ unit: (newSelectedOption || options[0]).value })
    }, []),
    selectedOption = React.useMemo(() => options.find(option => option.value === props.value), [props.value, options])

  return (
    <Select
      name="field-lengthCSS-select"
      isClearable
      isSearchable
      placeholder='place'
      value={selectedOption}
      options={options}
      onChange={onChange}
      menuPosition="fixed"
      {...(process.env.IS_TESTS && { classNamePrefix: 'select' })}
    />
  )
})
