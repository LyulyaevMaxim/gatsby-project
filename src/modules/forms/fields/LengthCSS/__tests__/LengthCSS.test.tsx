import React from 'react'
import { render, fireEvent, RenderResult } from '@testing-library/react'
// import cases from "jest-in-case";
import { createSelectInterface } from 'modules/forms/select/testsHelper'
import * as NFieldLengthCSS from '../@types'
import { FieldLengthCSS } from '../index'
import { FieldInputLengthCSS, patterns, examplesForInputPatterns } from '../input'
import { FieldSelectLengthCSS } from '../select'

jest.mock('../index', () => require('../LengthCSS'))

const keysUnitsOfMeasure = {
    //TODO: way getting keys and values need to combine
    default: Object.keys(NFieldLengthCSS.defaultUnitsOfMeasure),
    length: Object.keys(NFieldLengthCSS.unitsOfLengthMeasure),
    additional: Object.keys(NFieldLengthCSS.additionalUnitsOfMeasure),
  },
  valuesUnitsOfMeasure = {
    default: Object.values(NFieldLengthCSS.defaultUnitsOfMeasure),
    length: Object.values(NFieldLengthCSS.unitsOfLengthMeasure),
    additional: Object.values(NFieldLengthCSS.additionalUnitsOfMeasure),
  }

describe('FieldLengthCSS', () => {
  const menuItems = ['auto', 'px', '%', 'rem'] as Array<NFieldLengthCSS.allUnits>,
    someNumber = '-12.678',
    someDefaultValue = {
      key: keysUnitsOfMeasure.default[0],
      value: valuesUnitsOfMeasure.default[0], //NFieldLengthCSS.defaultUnitsOfMeasure[keysUnitsOfMeasure.default[0]]
    },
    someLengthValue = keysUnitsOfMeasure.length[0]

  it('renders correctly', () => {
    const Field = render(<FieldLengthCSS menu-items={menuItems} value="20vw" />)
    expect(Field.asFragment()).toMatchSnapshot()
  })

  it(`props "menu-items" must change amount displaying options select`, async () => {
    const Field = render(<FieldLengthCSS />),
      Select = createSelectInterface({ container: Field.container })
    expect(Select.options.length).toBe(Object.keys(NFieldLengthCSS.allUnits).length)
    Field.rerender(<FieldLengthCSS menu-items={menuItems} />)
    expect(Select.options.length).toBe(menuItems.length)
  })

  it(`"input" and "select" must correctly influence each other`, () => {
    const Field = render(<FieldLengthCSS value={someDefaultValue.key} menu-items={menuItems} />),
      Select = createSelectInterface({ container: Field.container }),
      Input = createInputInterface({ Field })

    /* when was input <number> then "defaultValue" of select must be replaced
       to the first numeric unit of measure from 'menuItems' */
    expect(Select.value).toBe(someDefaultValue.key)
    Input.changeOnBlur(someNumber)
    expect(Select.value).toBe(menuItems.find(option => Object.keys(NFieldLengthCSS.numericUnits).includes(option)))

    /* when was input <number><unit of measure> then the input must be replaced
       to <number> and the select to <unit> */
    Field.rerender(<FieldLengthCSS menu-items={menuItems} />)
    Input.changeOnBlur(`${someNumber}${someLengthValue}`)
    expect(Input.value).toBe(someNumber)
    expect(Select.value).toBe(someLengthValue)

    // when was input <default value> then the select must be replaced to this default value too
    Field.rerender(<FieldLengthCSS menu-items={menuItems} />)
    Input.changeOnBlur(someDefaultValue.key)
    expect(Input.value).toBe(someDefaultValue.key)
    expect(Select.value).toBe(someDefaultValue.key)

    // when select was set to <default value> and input had a non default value then input will set default value too
    Field.rerender(<FieldLengthCSS menu-items={menuItems} value={someNumber} />)
    Select.changeOption(someDefaultValue.value)
    expect(Input.value).toBe(someDefaultValue.key)

    // when select was set to <numeric value> and input has a default value then input will reset
    Field.rerender(<FieldLengthCSS menu-items={menuItems} value={someDefaultValue.key} />)
    Select.changeOption('px') //TODO: need get menuItems element which in valuesUnitsOfMeasure.length
    expect(Input.value).toBe('')
  })
})

describe('FieldInputLengthCSS', () => {
  it('during input need reset validation status and only valid value must be pass', () => {
    const Input = createInputInterface({ Field: render(<FieldInputLengthCSS value="" />) })
    expect(Input.isValid).toBe('null')
    ;[
      { newInputValue: examplesForInputPatterns.invalidForInput[0], toBeValue: '' },
      {
        newInputValue: examplesForInputPatterns.validForInput[0],
        toBeValue: examplesForInputPatterns.validForInput[0],
      },
    ].forEach(({ newInputValue, toBeValue }) => {
      Input.changeOnInput(newInputValue)
      expect(Input.value).toBe(toBeValue)
      expect(Input.isValid).toBe('null')
    })
  })

  it('when saving value must be invalid or valid', () => {
    const Input = createInputInterface({ Field: render(<FieldInputLengthCSS value="" />) })
    expect(Input.isValid).toBe('null')
    ;[
      { newInputValue: examplesForInputPatterns.invalidForSaving[0], toBeValue: 'false' },
      { newInputValue: examplesForInputPatterns.validForSaving[0], toBeValue: 'true' },
    ].forEach(({ newInputValue, toBeValue }) => {
      Input.changeOnBlur(newInputValue)
      expect(Input.isValid).toBe(toBeValue)
    })
  })
})

describe('Input patterns', () => {
  const { similarInput, correctInput } = patterns

  it('valid data at the time of input', () =>
    examplesForInputPatterns.validForInput.forEach(inputtedValue =>
      expect(similarInput.test(inputtedValue)).toBeTruthy()
    ))

  it('invalid data at the time of input', () =>
    examplesForInputPatterns.invalidForInput.forEach(inputtedValue =>
      expect(similarInput.test(inputtedValue)).toBeFalsy()
    ))

  it('valid data at the time of saving', () =>
    examplesForInputPatterns.validForSaving.forEach(inputtedValue =>
      expect(correctInput.test(inputtedValue)).toBeTruthy()
    ))

  it('invalid data at the time of saving', () =>
    examplesForInputPatterns.invalidForSaving.forEach(inputtedValue =>
      expect(correctInput.test(inputtedValue)).toBeFalsy()
    ))
})

describe('FieldSelectLengthCSS', () => {
  const createSelect = (props: NFieldLengthCSS.IFieldSelectLengthCSS) => render(<FieldSelectLengthCSS {...props} />),
    menuItems = [keysUnitsOfMeasure.length[0], keysUnitsOfMeasure.default[0]] as Array<NFieldLengthCSS.allUnits>,
    nonExistValue = keysUnitsOfMeasure.additional[0] as NFieldLengthCSS.allUnits

  it(`if 'value' does not include in 'menuItems' then it added in options select`, () => {
    const Select = createSelectInterface({ container: createSelect({ menuItems, value: nonExistValue }).container })
    expect(Select.options.length).toBe(menuItems.length + 1)
    expect([...Select.options].map(option => option.textContent).includes(nonExistValue)).toBeTruthy()
  })

  it(`when 'value', which not includes in 'menuItems' change to any other, then the previous value delete from options select`, async () => {
    const Field = createSelect({ menuItems, value: nonExistValue }),
      Select = createSelectInterface({ container: Field.container })
    expect(Select.options.length).toBe(menuItems.length + 1)
    Field.rerender(<FieldSelectLengthCSS {...{ menuItems, value: menuItems[0] }} />)
    expect(Select.options.length).toBe(menuItems.length)
  })

  it(`when change to value, which not includes in 'menuItems' this value add in options select`, async () => {
    const Field = createSelect({ menuItems, value: menuItems[0] }),
      Select = createSelectInterface({ container: Field.container })

    expect(Select.options.length).toBe(menuItems.length)
    Field.rerender(<FieldSelectLengthCSS {...{ menuItems, value: nonExistValue }} />)
    expect(Select.options.length).toBe(menuItems.length + 1)
  })
})

interface ICreateInputInterface {
  Field: RenderResult
}

function createInputInterface({ Field }: ICreateInputInterface) {
  return {
    get self() {
      return Field.getByTestId('field-lengthCSS-input') as any
    },
    get value() {
      return this.self.value
    },
    get isValid() {
      return this.self.dataset.isValid
    },
    changeOnBlur(newInputValue: string) {
      fireEvent.blur(this.self, { target: { value: newInputValue } })
    },
    changeOnInput(newInputValue: string) {
      fireEvent.input(this.self, { target: { value: newInputValue } })
    },
  }
}
