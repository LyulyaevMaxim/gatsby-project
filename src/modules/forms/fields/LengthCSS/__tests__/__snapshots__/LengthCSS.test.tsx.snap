// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldLengthCSS renders correctly 1`] = `
<DocumentFragment>
  .emotion-10 {
  position: relative;
  box-sizing: border-box;
}

.emotion-9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: hsl(0,0%,100%);
  border-color: hsl(0,0%,80%);
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: default;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  min-height: 38px;
  outline: 0 !important;
  position: relative;
  -webkit-transition: all 100ms;
  transition: all 100ms;
  box-sizing: border-box;
}

.emotion-9:hover {
  border-color: hsl(0,0%,70%);
}

.emotion-2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  padding: 2px 8px;
  -webkit-overflow-scrolling: touch;
  position: relative;
  overflow: hidden;
  box-sizing: border-box;
}

.emotion-0 {
  color: hsl(0,0%,20%);
  margin-left: 2px;
  margin-right: 2px;
  max-width: calc(100% - 8px);
  overflow: hidden;
  position: absolute;
  text-overflow: ellipsis;
  white-space: nowrap;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  box-sizing: border-box;
}

.emotion-1 {
  margin: 2px;
  padding-bottom: 2px;
  padding-top: 2px;
  visibility: visible;
  color: hsl(0,0%,20%);
  box-sizing: border-box;
}

.emotion-8 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  box-sizing: border-box;
}

.emotion-4 {
  color: hsl(0,0%,80%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 8px;
  -webkit-transition: color 150ms;
  transition: color 150ms;
  box-sizing: border-box;
}

.emotion-4:hover {
  color: hsl(0,0%,60%);
}

.emotion-3 {
  display: inline-block;
  fill: currentColor;
  line-height: 1;
  stroke: currentColor;
  stroke-width: 0;
}

.emotion-5 {
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  background-color: hsl(0,0%,80%);
  margin-bottom: 8px;
  margin-top: 8px;
  width: 1px;
  box-sizing: border-box;
}

<div
    css="[object Object]"
  >
    <input
      css="[object Object],[object Object]"
      data-is-valid="true"
      data-testid="field-lengthCSS-input"
      type="text"
      value="20"
    />
    <div
      class=" emotion-10"
    >
      <div
        class="select__control emotion-9"
      >
        <div
          class="select__value-container select__value-container--has-value emotion-2"
        >
          <div
            class="select__single-value emotion-0"
          >
            vw
          </div>
          <div
            class="emotion-1"
          >
            <div
              class="select__input"
              style="display: inline-block;"
            >
              <input
                aria-autocomplete="list"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                id="react-select-2-input"
                spellcheck="false"
                style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
              <div
                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
              />
            </div>
          </div>
        </div>
        <div
          class="select__indicators emotion-8"
        >
          <div
            aria-hidden="true"
            class="select__indicator select__clear-indicator emotion-4"
          >
            <svg
              aria-hidden="true"
              class="emotion-3"
              focusable="false"
              height="20"
              viewBox="0 0 20 20"
              width="20"
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </div>
          <span
            class="select__indicator-separator emotion-5"
          />
          <div
            aria-hidden="true"
            class="select__indicator select__dropdown-indicator emotion-4"
          >
            <svg
              aria-hidden="true"
              class="emotion-3"
              focusable="false"
              height="20"
              viewBox="0 0 20 20"
              width="20"
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
      <input
        name="field-lengthCSS-select"
        type="hidden"
        value="vw"
      />
    </div>
  </div>
</DocumentFragment>
`;
